File: search_and_upload_composio.py
import requests
import os

# Composio config
COMPOSIO_API_KEY = os.environ.get("COMPOSIO_API_KEY", "").strip()
DRIVE_FOLDER_ID = "1xoOfpTUwxMa_pIHoP71aCDH0Eb03tzyf"
GOOGLE_CREDENTIALS_PATH = "/Users/jeremy/dev/Meine-Google-Credentials/credentials.json"

# Search config
CONTRACT_KEYWORDS = [
    "Vertrag", "Vereinbarung", "Mitgliedschaft", "Subscription", 
    "Abonnement", "Kaufvertrag", "Mietvertrag", "Arbeitsvertrag",
    "Versicherung", "Darlehen", "Kredit", "Leasing"
]
EXCLUDE_KEYWORDS = ["rechnung", "beleg", "invoice", "receipt", "quittung"]

def get_gmail_accounts():
    resp = requests.get(
        "https://backend.composio.dev/api/v3/connected_accounts?status=ACTIVE",
        headers={"x-api-key": COMPOSIO_API_KEY},
    )
    items = resp.json().get("items", [])
    return [acc for acc in items if "gmail" in acc.get("toolkit", {}).get("slug", "")]

def execute_action(action, params, account_id):
    url = "https://backend.composio.dev/api/v1/execute"
    payload = {
        "action": action,
        "parameters": params,
        "connectedAccountId": account_id
    }
    resp = requests.post(url, headers={"x-api-key": COMPOSIO_API_KEY}, json=payload)
    if resp.status_code != 200:
        print(f"  Error {resp.status_code}: {resp.text}")
        return None
    return resp.json()

def main():
    if not COMPOSIO_API_KEY:
        raise SystemExit("Missing COMPOSIO_API_KEY environment variable")

    accounts = get_gmail_accounts()
    print(f"Found {len(accounts)} active Gmail accounts in Composio.")

    for acc in accounts:
        acc_id = acc["id"]
        email = acc.get("connectionParams", {}).get("email", acc_id)
        print(f"\nProcessing account: {email}")

        for kw in CONTRACT_KEYWORDS:
            query = f"{kw} " + " ".join([f"-{ex}" for ex in EXCLUDE_KEYWORDS]) + " has:attachment"
            print(f"  Searching for: {kw}")
            
            res = execute_action("GMAIL_FETCH_EMAILS", {"query": query, "maxResults": 10}, acc_id)
            if not res: continue
            
            # The structure of res['output'] can vary. Based on typical Composio responses:
            messages = res.get("output", {}).get("messages", [])
            if not messages and "data" in res.get("output", {}):
                messages = res["output"]["data"]
            
            if not isinstance(messages, list):
                print(f"    Unexpected response format: {type(messages)}")
                continue

            print(f"    Found {len(messages)} potential contract emails.")

            for msg in messages:
                msg_id = msg.get("id")
                # To get attachments, we usually need the message details or use a specific tool
                # Let's try to get details for each message
                details = execute_action("GMAIL_GET_MAIL", {"messageId": msg_id}, acc_id)
                if not details: continue
                
                # Extract attachments from details
                payload = details.get("output", {}).get("payload", {})
                parts = payload.get("parts", [])
                
                def find_attachments(parts_list, found=None):
                    if found is None: found = []
                    for p in parts_list:
                        if p.get("filename") and p.get("body", {}).get("attachmentId"):
                            found.append(p)
                        if p.get("parts"):
                            find_attachments(p["parts"], found)
                    return found

                attachments = find_attachments(parts)
                for att in attachments:
                    fname = att["filename"]
                    att_id = att["body"]["attachmentId"]
                    print(f"      Downloading attachment: {fname}")
                    
                    # Download attachment
                    att_res = execute_action("GMAIL_GET_ATTACHMENT", {
                        "messageId": msg_id,
                        "attachmentId": att_id
                    }, acc_id)
                    
                    if att_res and "output" in att_res:
                        # Upload to Drive using the Drive logic from earlier (but in Python)
                        # For now, let's just log the success if we have the data
                        # Actually, I should use the Google Drive API directly here to be sure.
                        print(f"      Ready to upload {fname} to Drive folder {DRIVE_FOLDER_ID}")
                        # I will implement the actual upload in the next step when I have the data
                        pass

if __name__ == "__main__":
    main()

File: search_and_upload_composio.ts
import axios from 'axios';
import { google } from 'googleapis';
import { JWT } from 'google-auth-library';
import * as fs from 'fs';
import * as path from 'path';

const COMPOSIO_API_KEY = "ak_d7C_5oCKr5NS0gDyANN-";
const DRIVE_FOLDER_ID = "1xoOfpTUwxMa_pIHoP71aCDH0Eb03tzyf";
const GOOGLE_CREDENTIALS_PATH = "/Users/jeremy/dev/Meine-Google-Credentials/credentials.json";

const CONTRACT_KEYWORDS = [
    "Vertrag", "Vereinbarung", "Mitgliedschaft", "Subscription", 
    "Abonnement", "Kaufvertrag", "Mietvertrag", "Versicherung"
];
const EXCLUDE_KEYWORDS = ["rechnung", "beleg", "invoice", "receipt", "quittung"];

async function main() {
    const auth = new JWT({
        keyFile: GOOGLE_CREDENTIALS_PATH,
        scopes: ['https://www.googleapis.com/auth/drive.file']
    });
    const drive = google.drive({ version: 'v3', auth });

    console.log("Fetching connected Gmail accounts from Composio...");
    const accountsResp = await axios.get("https://backend.composio.dev/api/v3/connected_accounts?status=ACTIVE", {
        headers: { "x-api-key": COMPOSIO_API_KEY }
    });
    const gmailAccounts = accountsResp.data.items.filter((acc: any) => acc.toolkit.slug.includes("gmail"));

    console.log(`Found ${gmailAccounts.length} active Gmail accounts.`);

    for (const acc of gmailAccounts) {
        console.log(`\nSearching account: ${acc.id}`);
        
        for (const kw of CONTRACT_KEYWORDS) {
            const query = `${kw} ${EXCLUDE_KEYWORDS.map(ex => `-${ex}`).join(" ")} has:attachment`;
            console.log(`  Query: ${query}`);
            
            try {
                const fetchResp = await axios.post("https://backend.composio.dev/api/v1/execute", {
                    action: "gmail_fetch_emails", // lowercase attempt
                    parameters: { query, max_results: 10 },
                    connectedAccountId: acc.id
                }, { headers: { "x-api-key": COMPOSIO_API_KEY } });

                const messages = fetchResp.data.output?.data || fetchResp.data.output?.messages || [];
                if (!Array.isArray(messages)) continue;

                for (const msg of messages) {
                    console.log(`    Processing message ID: ${msg.id}`);
                    
                    const detailsResp = await axios.post("https://backend.composio.dev/api/v1/execute", {
                        action: "gmail_get_mail",
                        parameters: { message_id: msg.id },
                        connectedAccountId: acc.id
                    }, { headers: { "x-api-key": COMPOSIO_API_KEY } });

                    const payload = detailsResp.data.output?.payload || {};
                    const parts = payload.parts || [];
                    
                    const findAttachments = (partsList: any[], found: any[] = []) => {
                        for (const p of partsList) {
                            if (p.filename && p.body?.attachmentId) found.push(p);
                            if (p.parts) findAttachments(p.parts, found);
                        }
                        return found;
                    };

                    const attachments = findAttachments(parts);
                    for (const att of attachments) {
                        console.log(`      Found attachment: ${att.filename}`);
                        
                        const attResp = await axios.post("https://backend.composio.dev/api/v1/execute", {
                            action: "gmail_get_attachment",
                            parameters: { message_id: msg.id, attachment_id: att.body.attachmentId },
                            connectedAccountId: acc.id
                        }, { headers: { "x-api-key": COMPOSIO_API_KEY } });

                        const base64Data = attResp.data.output?.data || attResp.data.output?.attachmentData;
                        if (base64Data) {
                            const tempPath = path.join("/tmp", att.filename);
                            fs.writeFileSync(tempPath, Buffer.from(base64Data, 'base64'));
                            
                            console.log(`      Uploading ${att.filename} to Drive...`);
                            await drive.files.create({
                                requestBody: { name: att.filename, parents: [DRIVE_FOLDER_ID] },
                                media: { mimeType: att.mimeType, body: fs.createReadStream(tempPath) }
                            });
                            fs.unlinkSync(tempPath);
                        }
                    }
                }
            } catch (err: any) {
                console.error(`  Error searching for ${kw}: ${err.message}`);
                if (err.response?.data) console.error("  Details:", err.response.data);
            }
        }
    }
}

main().catch(console.error);
